Step5 로또(수동) 코드 리뷰 요청드립니다.

dongha-byun opened the pull request:
안녕하세요.
우아한 테크캠프 Pro 5기 교육생 변동하 입니다.

로또(수동) 코드리뷰를 요청드리니 확인 부탁드립니다...!

이번 PR이 리뷰어님과 함께하는 PR이네요 ㅠ 
그동안 부족한 코드보고 섬세한 리뷰 해주시느라 너무 고생많으셨고 감사했습니다.
덕분에 1주차 미션 하는동안 정말 많은걸 배울 수 있었습니다.
본업하시면서 리뷰어 활동까지 하시느라 힘드실텐데, 요즘들어 날씨가 추워지니 몸조리 잘 하시길 바라겠습니다...!
그동안 감사했습니다.

dongha-byun committed:
Docs: 로또(수동) 기능 목록 정리

dongha-byun committed:
Test: 수동 로또 갯수 관련 테스트 코드 작성

dongha-byun committed:
Feat: 수동 로또 갯수 구현코드 작성

dongha-byun committed:
Docs: 기능목록 정리 수정

dongha-byun committed:
Test: 수동 로또 번호 입력 class test

dongha-byun committed:
Feat: 수동 로또 번호 class 구현코드 작성

dongha-byun committed:
Fix: 수동 로또 번호 범위 검증 추가

dongha-byun committed:
Refactor: 수동 로또 refactoring
- ManualLottoInput class 삭제
- ConsoleLottoNumberStrategy 검증로직 추가
- Input/Output view 수정

dongha-byun committed:
Fix: 입력 오류에 대한 예외처리 추가

dongha-byun committed:
Refactor: 미사용 import 제거

chae-yh commented on src/main/java/lotto/README.md:
```diff
@@ -59,3 +59,29 @@
 
+--------------------------------------------------------------
+
+# 3. 추가 기능 정리 - 로또(수동)
```
기능 요구 사항 정리 마지막까지 잘 작성해주셨습니다 👍

chae-yh commented on src/main/java/lotto/domain/BuyingLottoGroup.java:
```diff
@@ -8,8 +8,12 @@ public class BuyingLottoGroup {
     private List<Lotto> lottos;
 
     public static BuyingLottoGroup create(int count, LottoNumberStrategy lottoNumberStrategy) {
```
오버로딩을 통해서 구분을 해주었네요 ㅎㅎ 재밌는 방법인 것 같습니다 👍

chae-yh commented on src/main/java/lotto/domain/BuyingLottoGroup.java:
```diff
@@ -8,8 +8,12 @@ public class BuyingLottoGroup {
     private List<Lotto> lottos;
 
     public static BuyingLottoGroup create(int count, LottoNumberStrategy lottoNumberStrategy) {
+        return create(count, lottoNumberStrategy, new ArrayList<>());
```
빈 리스트를 전달한다면, Collections.emptyList() 는 어떨까요...?

dongha-byun replied:
오 그런 API도 있었군요...! 확인해보고 수정해보도록 하겠습니다!

chae-yh commented on src/main/java/lotto/strategy/ConsoleLottoNumberStrategy.java:
```diff
@@ -1,29 +1,76 @@
 
-    public void validateOnlyNumber(String input){
-        if(!Pattern.matches("^\\d*$", input)){
+    private boolean validateInput(List<String> numberStrings) {
```
이 부분은 어쩔 수 없는 부분일 수 있겠지만, 결국 Lotto의 validation 로직과 거의 동일해보이네요. 바로 Lotto를 만들어주면 중복 코드를 조금 더 줄일 수도 있을 것 같습니다..!
혹시 이 부분에 대한 로직을 Lotto 쪽으로 옮겨볼 수 있을까요...? 힘든 작업이라면 말씀주시고, pr 주시면 merge 하겠습니다 ㅎㅎ

dongha-byun replied:
저도 구현하면서 아 이건 중복이 좀 심한데 라는 생각을 했는데 각 도메인에서 검증을 하다보니 어쩔수없나 싶었던 부분이었습니다. 말씀하신 방안대로 고민 한번 해겠습니다!

chae-yh replied:
인터페이스에서 Lotto를 반환하도록 하면서 자동으로 validation이 되게 했네요 👍

chae-yh commented on src/test/java/lotto/domain/ManualLottoCountTest.java:
```diff
@@ -0,0 +1,32 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+public class ManualLottoCountTest {
```
수동 로또 개수에 대한 테스트도 좋습니다 👍

chae-yh commented on src/main/java/lotto/domain/ManualLottoCount.java:
```diff
@@ -0,0 +1,38 @@
+
+import java.util.regex.Pattern;
+
+public class ManualLottoCount {
```
수동 로또 개수 👍

chae-yh commented on src/main/java/lotto/service/LottoService.java:
```diff
@@ -23,17 +30,38 @@ public Lotto generateWinningBasicLotto() {
+        try{
+            return ManualLottoCount.create(buyManualLotto(), lottoPayment.getLottoCount());
+        }catch(IllegalArgumentException e){
+            System.err.println("[ERROR] "+e.getLocalizedMessage());
```
system 함수는 지양해주세요..! 리소스 사용 등에 이슈가 있을 수 있습니다.

dongha-byun replied:
넵! 수정하도록 하겠습니당

chae-yh commented on src/main/java/lotto/service/LottoService.java:
```diff
@@ -23,17 +30,38 @@ public Lotto generateWinningBasicLotto() {
+            return ManualLottoCount.create(buyManualLotto(), lottoPayment.getLottoCount());
+        }catch(IllegalArgumentException e){
+            System.err.println("[ERROR] "+e.getLocalizedMessage());
+            return generateManualLottoCount(lottoPayment);
```
에러가 날 때마다 다시 값을 받을 수 있도록 만드신 것 같은데, 좋네요 ㅎㅎ

chae-yh reviewed (changes requested):
동하님 안녕하세요. 벌써 마지막 step이라니 아쉬움까지 느껴지네요..ㅠ
기존에 구조가 잘 잡혀서인지 많은 수정은 없었지만 기능에 대한 요구 사항을 잘 만족시켜주는 것 같습니다 ㅎㅎ 바로 merge 해도 좋겠지만 조금만 고민해주시면 좋을만한 내용들이 있어 코멘트 드리니 확인 부탁드립니다.

감사합니다 : )

dongha-byun committed:
Refactor: System.err 제거 -> log.error 로 변경

dongha-byun committed:
Refactor: new ArrayList 대신 Collections.emptyList() 사용

dongha-byun committed:
Refactor: LottoNumbersStrategy class 수정 및 관련 코드 리펙터링
- 로또 생성 전략 코드 수정 ( LottoNumbersStrategy interface -> LottoGenerateStrategy interface)
- Lotto class validate 추가 (validateCount) 및 strategy 중복 제거

chae-yh reviewed (approved):
동하님 안녕하세요. 피드백 반영 잘해주셨습니다 👍 그리고, 더 코멘트 드릴 내용이 없네요 ㅎㅎ 이만 merge 하겠습니다.

제 나름대로 열심히 리뷰드렸다고 생각했는데, 동하님께도 그 마음이 전달되었는지 모르겠네요 ㅎㅎ 이제 제 리뷰는 여기까지지만, 남은 미션에서 저보다 뛰어나신 리뷰어님들과 더 많은 성장이루시길 바라겠습니다...!

그 동안 고생많으셨습니다 감사합니다 : )
